### PR TITLE
feat: reintroducing thumbnails

### PR DESCRIPTION
**Pull request by RoccoSmit**
PR to cover the Image Preview issue
https://github.com/usememos/memos/issues/3685

Design specs:
- Thumbnails should reside in the local file system as an optional cache.
- Only attempt to access the thumbnail when requested, and if it doesn't exist, create it and then return it.

Changes made:
- Updated api request signature to allow for a thumbnail query string variable
- Followed the logic from the previous resource flow [resource/resource.go](https://github.com/usememos/memos/commit/ff175bbb7e4ff8e5407786ec73bcb361033fd425#diff-7f38be2e91afd9cd1efefe248b880b8cfa4539236d228462dff7169a19ab114c) with these notable exceptions
  - The resource request will look for a thumbnail first and return it if found, rather than load the original image first then generate the thumbnail after. This limits the number of larger images being read from disk.
  - Newly generated thumbnail images are converted to byte arrays on the fly and returned, removing the need to write images to the disk and then reading them back to get the byte array.
- Moved thumbnail logic into its own file as it gets used across services
- Added thumbnail deleting logic in individual services. Resource files are deleted in the `store` logic, but no version of thumbnails have db related data so `store` logic did not feel like the place to put thumbnail file deletion. 
 
Notes:
- The first load of the home / resource pages after this feature is enabled will cause stress on the system as each image request will attempt to generate a thumbnail. Subsequent loads should be improved as thumbnails will load.
- These changes were manually tested with `sqlite` and `local storage` storage options and work as expected

### CODE REVIEW

File: server/router/api/v1/thumbnail.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	if err := imaging.Save(thumbnailImage, path); err != nil {
+		return errors.Wrap(err, "failed to save thumbnail file")
+	}
+
+	return nil
+}
+
+func (t Thumbnail) ImageToBlob(thumbnailImage image.Image) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (RoccoSmit) [reply server/router/api/v1/thumbnail.go]: change made in latest commit

File: server/router/api/v1/thumbnail.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+	buf := new(bytes.Buffer)
+	if err := imaging.Encode(buf, thumbnailImage, imgFormat); err != nil {
+		return nil, errors.Wrap(err, "failed to convert thumbnail image to bytes")
+	}
+
+	return buf.Bytes(), nil
+}
+
+func (t Thumbnail) DeleteFile(assetsFolderPath string) error {
```
[the remark this answers is not in the document]
Comment (RoccoSmit) [reply server/router/api/v1/thumbnail.go]: change made in latest commit

File: server/router/api/v1/thumbnail.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+	dstBlob, err := io.ReadAll(dstFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read thumbnail file")
+	}
+
+	return dstBlob, nil
+}
+
+func GenerateThumbnailImage(sourceBlob []byte) (image.Image, error) {
```
[the remark this answers is not in the document]
Comment (RoccoSmit) [reply server/router/api/v1/thumbnail.go]: change made in latest commit

File: server/router/api/v1/thumbnail.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+		atomic.AddInt32(&availableGeneratorAmount, 1)
+	}()
+
+	reader := bytes.NewReader(sourceBlob)
+	src, err := imaging.Decode(reader, imaging.AutoOrientation(true))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to decode thumbnail image")
+	}
+
+	thumbnailImage := imaging.Resize(src, 512, 0, imaging.Lanczos)
```
Comment (michelheusschen) [inline server/router/api/v1/thumbnail.go:88]: Thank you for working on this, but I have a couple of concerns:
1. Images in the web client have a maximum width of `700px` at 100% zoom. When the thumbnail width is limited to `512px`, the images are no longer full width. Is there a reason behind this number?
2. Images less than 512px wide are upscaled. Maybe omit thumbnail generation in that case?

Comment (RoccoSmit) [reply server/router/api/v1/thumbnail.go:88]: Good catch

the thumbnail generation code was a lift and shift from v0.21, this is where the 512px comes from. I had a look through the history and when the code was added in v0.13 the timeline images had a max width of 320px so it worked in that scenario. Over the following versions timeline images got larger and the thumbnail flag was remove from them and only the resource page images, where images are < 512px, kept the flag

I'll update the code based on your feedback

**Review by boojack — APPROVED** [review]

**Pull request merged**